src/adapter/src/coord/sequencer/inner/cluster.rs

```diff
@@ -428,4 +469,105 @@ impl Coordinator {
             ObjectType::Cluster,
         )))
     }
+
+    async fn wait_for_pending_replicas_hydrated(
+        &mut self,
+        session: &Session,
+        plan: AlterClusterPlan,
+        new_config: ClusterVariantManaged,
+        timeout_time: Instant,
+        on_timeout: OnTimeoutAction,
+        validity: PlanValidity,
+    ) -> Result<StageResult<Box<ClusterStage>>, AdapterError> {
+        // wait and re-signal wait for hydrated if not hydrated
+        let cluster = self.catalog.get_cluster(plan.id);
+        let pending_replicas = cluster
+            .replicas()
+            .filter_map(|r| {
+                if r.config.location.pending() {
+                    Some(r.replica_id.clone())
+                } else {
+                    None
+                }
+            })
+            .collect_vec();
+        // Check For timeout
+        if Instant::now() > timeout_time {
+            // Timed out handle timeout action
+            match on_timeout {
+                OnTimeoutAction::Rollback => {
+                    self.active_conns
+                        .get_mut(session.conn_id())
+                        .expect("There must be an active connection")
+                        .pending_cluster_alters
+                        .remove(&cluster.id);
+                    let pending_replica_drops: Vec<DropObjectInfo> = pending_replicas
+                        .iter()
+                        .map(|r| {
+                            DropObjectInfo::ClusterReplica((
+                                cluster.id.clone(),
+                                r.to_owned(),
+                                ReplicaCreateDropReason::Manual,
+                            ))
+                        })
+                        .collect::<Vec<DropObjectInfo>>();
+                    self.catalog_transact(None, vec![Op::DropObjects(pending_replica_drops)])
+                        .await?;
+                    return Err(AdapterError::AlterClusterTimeout);
```

Comment:
Isn't this logic duplicated somewhere else from a previous PR? I think we should extract it to it's own method.